File: app/src/main/java/com/example/videoplayer/ui/AudioTrackBottomSheet.java
package com.example.videoplayer.ui;

import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.lifecycle.Lifecycle;
import androidx.lifecycle.LiveData;
import androidx.lifecycle.Observer;
import androidx.lifecycle.ViewModelProvider;
import androidx.recyclerview.widget.LinearLayoutManager;

import com.example.videoplayer.databinding.BottomSheetAudioTracksBinding;
import com.example.videoplayer.model.AudioTrack;
import com.example.videoplayer.ui.adapter.AudioTrackAdapter;
import com.example.videoplayer.viewmodel.PlayerViewModel;
import com.google.android.material.bottomsheet.BottomSheetDialogFragment;

import java.util.List;

/**
 * Material BottomSheetDialogFragment that lists all available audio tracks
 * and allows the user to switch between them with a single tap.
 */
public class AudioTrackBottomSheet extends BottomSheetDialogFragment {

    private BottomSheetAudioTracksBinding binding;

    // Shared ViewModel – same instance as PlayerActivity's
    private PlayerViewModel viewModel;

    private final AudioTrackAdapter adapter = new AudioTrackAdapter(track -> onTrackSelected(track));

    public static AudioTrackBottomSheet newInstance() {
        return new AudioTrackBottomSheet();
    }

    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container,
                             @Nullable Bundle savedInstanceState) {
        binding = BottomSheetAudioTracksBinding.inflate(inflater, container, false);
        return binding.getRoot();
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);

        // Get ViewModel from the parent Activity using ViewModelProvider (generic, no coupling)
        if (getActivity() != null) {
            viewModel = new ViewModelProvider(getActivity()).get(PlayerViewModel.class);
        }

        if (viewModel == null) {
            return;
        }

        // Set up RecyclerView
        binding.rvAudioTracks.setLayoutManager(new LinearLayoutManager(requireContext()));
        binding.rvAudioTracks.setAdapter(adapter);

        // Observe audio tracks reactively
        final Observer<List<AudioTrack>> tracksObserver = tracks -> {
            if (tracks == null || tracks.isEmpty()) {
                binding.tvNoTracks.setVisibility(View.VISIBLE);
                binding.rvAudioTracks.setVisibility(View.GONE);
            } else {
                binding.tvNoTracks.setVisibility(View.GONE);
                binding.rvAudioTracks.setVisibility(View.VISIBLE);
                adapter.submitList(tracks);
            }
        };

        viewModel.audioTracks.observe(getViewLifecycleOwner(), tracksObserver);

        // Observe selected track index to keep the check mark in sync
        final Observer<Integer> indexObserver = index -> {
            if (index != null) {
                adapter.setSelectedIndex(index);
            }
        };

        viewModel.playerManager.selectedAudioIndex.observe(getViewLifecycleOwner(), indexObserver);
    }

    /**
     * Called when the user taps an audio track row.
     * Delegates to ViewModel which routes to PlayerManager -> TrackSelectionOverride.
     */
    private void onTrackSelected(AudioTrack track) {
        if (viewModel != null) {
            viewModel.selectAudioTrack(track);
        }
        dismiss();
    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();
        binding = null;
    }

    public static final String TAG = "AudioTrackBottomSheet";
}

File: app/src/main/java/com/example/videoplayer/ui/AudioTrackBottomSheet.kt
package com.example.videoplayer.ui

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.fragment.app.activityViewModels
import androidx.lifecycle.Lifecycle
import androidx.lifecycle.lifecycleScope
import androidx.lifecycle.repeatOnLifecycle
import androidx.recyclerview.widget.LinearLayoutManager
import com.example.videoplayer.databinding.BottomSheetAudioTracksBinding
import com.example.videoplayer.model.AudioTrack
import com.example.videoplayer.ui.adapter.AudioTrackAdapter
import com.example.videoplayer.viewmodel.PlayerViewModel
import com.google.android.material.bottomsheet.BottomSheetDialogFragment
import kotlinx.coroutines.launch

/**
 * Material BottomSheetDialogFragment that lists all available audio tracks
 * and allows the user to switch between them with a single tap.
 *
 * Communication with [PlayerActivity] happens through the shared [PlayerViewModel]
 * obtained via [activityViewModels] – no interface callbacks needed.
 */
class AudioTrackBottomSheet : BottomSheetDialogFragment() {

    private var _binding: BottomSheetAudioTracksBinding? = null
    private val binding get() = _binding!!

    // Shared ViewModel – same instance as PlayerActivity's
    private val viewModel: PlayerViewModel by activityViewModels()

    private val adapter = AudioTrackAdapter { track ->
        onTrackSelected(track)
    }

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View {
        _binding = BottomSheetAudioTracksBinding.inflate(inflater, container, false)
        return binding.root
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        // Set up RecyclerView
        binding.rvAudioTracks.layoutManager = LinearLayoutManager(requireContext())
        binding.rvAudioTracks.adapter = adapter

        // Observe audio tracks reactively
        viewLifecycleOwner.lifecycleScope.launch {
            repeatOnLifecycle(Lifecycle.State.STARTED) {
                viewModel.audioTracks.collect { tracks ->
                    if (tracks.isEmpty()) {
                        binding.tvNoTracks.visibility = View.VISIBLE
                        binding.rvAudioTracks.visibility = View.GONE
                    } else {
                        binding.tvNoTracks.visibility = View.GONE
                        binding.rvAudioTracks.visibility = View.VISIBLE
                        adapter.submitList(tracks)
                    }
                }
            }
        }

        // Observe selected track index to keep the check mark in sync
        viewLifecycleOwner.lifecycleScope.launch {
            repeatOnLifecycle(Lifecycle.State.STARTED) {
                viewModel.playerManager.selectedAudioIndex.collect { index ->
                    adapter.selectedIndex = index
                }
            }
        }
    }

    /**
     * Called when the user taps an audio track row.
     * Delegates to ViewModel which routes to PlayerManager → TrackSelectionOverride.
     */
    private fun onTrackSelected(track: AudioTrack) {
        viewModel.selectAudioTrack(track)
        // Dismiss after selection for a clean UX
        dismiss()
    }

    override fun onDestroyView() {
        super.onDestroyView()
        _binding = null
    }

    companion object {
        const val TAG = "AudioTrackBottomSheet"

        fun newInstance() = AudioTrackBottomSheet()
    }
}
